src/transformStateWithClones.js

```diff
@@ -7,7 +7,52 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  let previousState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        previousState = createStateWithAddedProperties(
+          previousState,
+          action.extraData,
+        );
+        break;
+      case 'removeProperties':
+        previousState = createStateWithRemovedProperties(
+          previousState,
+          action.keysToRemove,
```

Comment:
Similarly, the function name `createStateWithRemovedProperties` does not communicate that it creates a copy of the state with some properties removed. A better name could be `copyStateWithRemovedProperties`.